src/transformers/integrations/tensor_parallel.py

```diff
@@ -139,15 +140,81 @@ def get_packed_weights(param, empty_param, device_mesh, rank, dim):
+    rank 7 gets					(2, 5120, 8190)			 (14 ... 16, 5120, 8190)
+    rank 8 gets					(0, 5120, 8190)
+    rank 9 gets					(0, 5120, 8190)
+    rank 10 gets			    (0, 5120, 8190)
+    rank 11 gets				(0, 5120, 8190)
+    rank 12 gets				(0, 5120, 8190)
+    rank 13 gets				(0, 5120, 8190)
+
+    Case (3)
+    empty_param                 (16, 5120, 8190)
+    dim                         0
+    device_mesh.size()          3
+    rank 0 gets					(6, 5120, 8190)			 (0 ... 6, 5120, 8190)
+    rank 1 gets					(6, 5120, 8190)			 (6 ... 12, 5120, 8190)
+    rank 2 gets					(4, 5120, 8190)			 (12 ... 16, 5120, 8190)
+
+    In case (2), empty shards are returned with appropriate dimension to allow for operations to work smoothly.
+    """
+    if not (-len(empty_param.shape) <= dim < len(empty_param.shape)):
+        raise ValueError(
+            f"provided index {dim} should be within the range of number of parameter's dimensions {len(empty_param.shape)}"
+        )
+    _dim_size = empty_param.shape[dim]
+    _split_size = math.ceil(_dim_size / device_mesh.size())
     if dim == 0:
-        size_ = empty_param.shape[0]
-        param = param[rank * (size_ // device_mesh.size()) : (rank + 1) * (size_ // device_mesh.size()), ...]
-    elif dim == 1 or dim == -2:
-        size_ = empty_param.shape[-2]
-        param = param[..., rank * (size_ // device_mesh.size()) : (rank + 1) * (size_ // device_mesh.size()), :]
-    elif dim == 2 or dim == -1:
-        size_ = empty_param.shape[-1]
-        param = param[..., rank * (size_ // device_mesh.size()) : (rank + 1) * (size_ // device_mesh.size())]
+        if (rank * _split_size) < _dim_size:
+            param = param[rank * _split_size : min((rank + 1) * _split_size, _dim_size), ...]
+        else:
+            param = torch.empty((0, *empty_param.shape[1:]), dtype=torch.int64)
+    elif dim == 1:
+        if (rank * _split_size) < _dim_size:
+            param = param[:, rank * _split_size : min((rank + 1) * _split_size, _dim_size), ...]
+        else:
+            param = torch.empty((empty_param.shape[0], 0, empty_param.shape[2:]), dtype=torch.int64)
+    elif dim == -2:
+        if (rank * _split_size) < _dim_size:
+            param = param[..., rank * _split_size : min((rank + 1) * _split_size, _dim_size), :]
+        else:
+            param = torch.empty((*empty_param.shape[:-2], 0, empty_param.shape[-1]), dtype=torch.int64)
+    elif dim == 2:
+        if (rank * _split_size) < _dim_size:
+            param = param[:, :, rank * _split_size : min((rank + 1) * _split_size, _dim_size), ...]
+        else:
+            param = torch.empty((*empty_param.shape[:2], 0, *empty_param.shape[3:]), dtype=torch.int64)
+    elif dim == -1:
+        if (rank * _split_size) < _dim_size:
+            param = param[..., rank * _split_size : min((rank + 1) * _split_size, _dim_size)]
+        else:
+            param = torch.empty((*empty_param.shape[:-1], 0), dtype=torch.int64)
```

Comment:
thanks and sorry was out for 2 weeks!
I think this can be refactored to me more readable no?! this is the only thing missing for me!

Reply:
@ArthurZucker 
No worries, glad you took a break! Thanks for returning to this PR.

> I think this can be refactored to me more readable no?! this is the only thing missing for me!

Do you have any specific suggestions? Do you want me to use better variable names? and more assignments like keeping `rank * _split_size` in a variable and reuse something of that sort? 

Let me know, thank you.

Reply:
@ArthurZucker awaiting for you comments!

Reply:
sure @ArthurZucker please let me know, thanks

Reply:
Merged!

Reply:
Thanks working on it!